### PR TITLE
digester: retry on docker.GetDigest failures

### PR DESCRIPTION
**Pull request by tiraboschi**
Sometimes the image digester tool fails due to
`too many requests to registry` error from the server
and this leads to unrecoverable error in the
Build and Push Images job.
This PR introduces a retry mechanism
to make it more resilient.

Signed-off-by: Simone Tiraboschi <stirabos@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tools/digester/main.go [hunk 1]
```diff
@@ -274,13 +276,29 @@ func readOneDigest(ctx context.Context, image *Image, index int, wg *sync.WaitGr
...
+			fmt.Println("retrying after error:", err)
+			time.Sleep(sleep)
+			sleep *= 2
+		}
+		digest, err = docker.GetDigest(ctx, sys, imgRef)
+		if err == nil {
+			return digest, nil
+		}
+	}
+	return "", fmt.Errorf("aborting after %d attempts, last error: %s", attempts, err)
```
Comment (nunnatsa) [inline tools/digester/main.go:299]: Please change `%s` to `%w`

```suggestion
	return "", fmt.Errorf("aborting after %d attempts, last error: %w", attempts, err)
```

File: tools/digester/main.go [hunk 2]
```diff
@@ -274,13 +276,29 @@ func readOneDigest(ctx context.Context, image *Image, index int, wg *sync.WaitGr
...
-	digest, err := docker.GetDigest(ctx, nil, imgRef)
+	digest, err := retryGetDigest(5, 1, ctx, nil, imgRef)
+
 	exitOnError(err, "failed to get digest from image")
 
 	ch <- message{index: index, digest: digest.Hex(), fullName: fullName}
 	wg.Done()
 }
 
+func retryGetDigest(attempts int, sleep time.Duration, ctx context.Context, sys *types.SystemContext, imgRef types.ImageReference) (digest digest.Digest, err error) {
```
Comment (nunnatsa) [inline tools/digester/main.go:287]: please move ctx to be the first parameter
```suggestion
func retryGetDigest(ctx context.Context, attempts int, sleep time.Duration, sys *types.SystemContext, imgRef types.ImageReference) (digest digest.Digest, err error) {
```

File: tools/digester/main.go [hunk 3]
```diff
@@ -274,13 +276,29 @@ func readOneDigest(ctx context.Context, image *Image, index int, wg *sync.WaitGr
 	imgRef, err := docker.ParseReference(fullName)
 	exitOnError(err, "failed to parse container reference")
 
-	digest, err := docker.GetDigest(ctx, nil, imgRef)
+	digest, err := retryGetDigest(5, 1, ctx, nil, imgRef)
```
Comment (nunnatsa) [inline tools/digester/main.go:279]: Duration of 1 is one nanosecond - isn't it too fast? maybe start from 1 second? Just need to make sure that the final duration won't cause timeout.

File: tools/digester/main.go [hunk 4]
```diff
@@ -25,6 +25,8 @@ import (
 	"encoding/csv"
 	"flag"
 	"fmt"
+	"github.com/containers/image/v5/types"
+	"github.com/opencontainers/go-digest"
```
Comment (nunnatsa) [inline tools/digester/main.go:29]: please move down - to not be with the standard library packages.

**Review by nunnatsa — COMMENTED** [review]
Added several comments. Most of them are cosmetic. But please consider to change the initial sleep time.

**Review by nunnatsa — APPROVED** [review]
/approve

File: tools/digester/main.go (hunk 1 above)
Comment (tiraboschi) [reply tools/digester/main.go]: done, thanks

File: tools/digester/main.go (hunk 2 above)
Comment (tiraboschi) [reply tools/digester/main.go]: done, thanks

File: tools/digester/main.go (hunk 4 above)
Comment (tiraboschi) [reply tools/digester/main.go]: done, thanks

File: tools/digester/main.go (hunk 3 above)
Comment (tiraboschi) [reply tools/digester/main.go]: Now it can potentially range from:
0, 1, 3, 9, 27 = 40 seconds up to:
0, 1+0.5, 3+1.5, 9+4.5, 27+13.5 = 60 seconds

**Pull request merged**